### DISCUSSION TITLE
Get-Types

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [SamErde/PSPreworkout#7] [General]
Here's another command that could benefit from a name tweak. `Get-Types` is rather generic, and that's not what the command does. The command gets *type accelerators* not a list of .NET types. Maybe something like `Get-PSTypeAccelerator` would be more meaningful?

**Comment by SamErde** [comment]
Should I cast the output as something other than the dictionary (System.Collections.Generic.Dictionary`2) type that it gives me?

**Reply by jdhitsolutions** [reply]
I am always looking for ways to write *rich* objects to the pipeline.

```powershell
   ([PSObject].Assembly.GetType('System.Management.Automation.TypeAccelerators')::Get).GetEnumerator() |
    ForEach-Object {
        #create a custom object to store the type name and the type itself
        [PSCustomObject]@{
            PSTypeName = 'PSTypeAccelerator'
            PSVersion  = $PSVersionTable.PSVersion
            Name       = $_.Key
            Type       = $_.Value.FullName
        }
    }
```

Using a defined type name means you can create a custom format file. You can use [New-PSFormatXML](http://bit.ly/31SGo5o) to create it.  I added the `PSVersion` property because there are differences between Windows PowerShell and PowerShell 7. I also figured the output doesn't need the full type object, just the full type name.

![image](https://github.com/user-attachments/assets/6f6f52f7-4d09-49cf-8bca-fbc8bade6664)

**Comment by SamErde** [comment]
Yes, that cleans up the output nicely and learning how to use PSFormatXML has been on my to-do list for a while; I just needed a good use case!

**Comment by jdhitsolutions** [comment]
In lieu of a pull request:

```powershell
function Get-PSTypeAccelerator {
    <#
    .SYNOPSIS
    List all available .NET types and type accelerators.

    .DESCRIPTION
    List all available types and type accelerators. These can be useful when trying to find or remember the different types and accellerators that are available to use in PowerShell.

    .EXAMPLE
    Get-Types

    #>
    [CmdletBinding()]
    [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseSingularNouns', 'Get-Types', Justification = 'The types are plural.')]
    param (
        [Parameter(Position = 0, HelpMessage = 'The name of the type accelerator like ADSI.')]
        [SupportsWildcards()]
        [string]$Name = "*"
    )

    $Get = ([PSObject].Assembly.GetType('System.Management.Automation.TypeAccelerators')::Get)

    $Get.GetEnumerator() | Where-Object {$_.key -like $Name} |
    ForEach-Object {
        #create a custom object to store the type name and the type itself
        [PSCustomObject]@{
            PSTypeName = 'PSTypeAccelerator'
            PSVersion  = $PSVersionTable.PSVersion
            Name       = $_.Key
            Type       = $_.Value.FullName
        }
    }

}
````

**Comment by SamErde** [comment]
Oh, sweet. I already got most of this updated but will double check and add your special sauce. 

Side note: for some reason I thought the HelpMessage parameter attribute wasn't available in WPS 5.1! I'm glad to be wrong about that. 👍

I also didn't know about `SupportsWildcards()`!

**Reply by jdhitsolutions** [reply]
`HelpMessage` for a parameter has been around a long time. Don't confuse this with the `ErrorMessage` attribute in `ValidateScript()`

```powershell
[ValidateScript({ [enum]::GetNames([Environment+SpecialFolder]) -contains $_ },
ErrorMessage = '{0} is an invalid special folder')]
```

This requires PowerShell 7.

**Reply by SamErde** [reply]
Ah, yep. That's what I was thinking of.
